app/uk/gov/hmrc/agentclientrelationships/util/RequestAwareLogging.scala

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright 2025 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.agentclientrelationships.util
+
+import play.api.Logger
+import play.api.http.HeaderNames
+import play.api.mvc.RequestHeader
+import uk.gov.hmrc.agentclientrelationships.util.RequestSupport._
+
+trait RequestAwareLogging {
+  val logger: RequestAwareLogger =
+    new RequestAwareLogger(
+      delegateLogger = Logger(getClass)
+    )
+}
+
+/** A logger which is aware of the request. It will append to the message extra information such as session ID, request ID, user agent, referer, and device ID
+  * etc.
+  *
+  * Logged messages are enriched with request-specific context.
+  */
+class RequestAwareLogger(
+  delegateLogger: Logger
+) {
+
+  def debug(message: => String)(implicit request: RequestHeader): Unit = logMessage(message, Debug)
+
+  def info(message: => String)(implicit request: RequestHeader): Unit = logMessage(message, Info)
+
+  def warn(message: => String)(implicit request: RequestHeader): Unit = logMessage(message, Warn)
+
+  def error(message: => String)(implicit request: RequestHeader): Unit = logMessage(message, Error)
+
+  def debug(
+    message: => String,
+    ex: Throwable
+  )(implicit request: RequestHeader): Unit = logMessage(
+    message,
+    ex,
+    Debug
+  )
+
+  def info(
+    message: => String,
+    ex: Throwable
+  )(implicit request: RequestHeader): Unit = logMessage(
+    message,
+    ex,
+    Info
+  )
+
+  def warn(
+    message: => String,
+    ex: Throwable
+  )(implicit request: RequestHeader): Unit = logMessage(
+    message,
+    ex,
+    Warn
+  )
+
+  def error(
+    message: => String,
+    ex: Throwable
+  )(implicit request: RequestHeader): Unit = logMessage(
+    message,
+    ex,
+    Error
+  )
+
+  private def context(implicit request: RequestHeader) = s"[context: ${request.method} ${request.path}] $sessionId $requestId $userAgent $referer $deviceId"
```

Comment:
May be worth raising with the wider team to decide on the list of fields we want